core/src/pow/types.rs

```diff
@@ -478,3 +499,47 @@ impl BitVec {
 		self.bits[pos / 8] |= 1 << (pos % 8) as u8;
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::ser::{BinReader, BinWriter, ProtocolVersion};
+	use rand::Rng;
+	use std::io::Cursor;
+
+	#[test]
+	fn test_proof_rw() {
+		for edge_bits in 10..64 {
+			let mut proof = Proof::new(gen_proof(edge_bits as u32));
+			proof.edge_bits = edge_bits;
+			let mut buf = Cursor::new(Vec::new());
+			let mut w = BinWriter::new(&mut buf, ProtocolVersion::local());
+			if let Err(e) = proof.write(&mut w) {
+				panic!("failed to write proof {:?}", e);
+			}
+			buf.set_position(0);
+			let mut r = BinReader::new(&mut buf, ProtocolVersion::local());
+			match Proof::read(&mut r) {
+				Err(e) => panic!("failed to read proof: {:?}", e),
+				Ok(p) => assert_eq!(p, proof),
+			}
+		}
+	}
+
+	fn gen_proof(bits: u32) -> Vec<u64> {
+		let mut rng = rand::thread_rng();
+		let mut v = Vec::with_capacity(42);
+		for _ in 0..42 {
+			v.push(rng.gen_range(
+				u64::pow(2, bits - 1),
```

Comment:
btw, isn't it easier to write 1 << (bits - 1) than u64::pow(2, bits - 1) ?

Reply:
I think it's easier to read u64::pow(2, bits - 1), so personally prefer it iff there's no performance hit.